src/Database/LSMTree.hs

```diff
@@ -0,0 +1,551 @@
+    checkBlobRefType i _ = throw (Internal.ErrBlobRefInvalid i)
+
+{-------------------------------------------------------------------------------
+  Snapshots
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE createSnapshot ::
+     (Common.Labellable (k, v, b), ResolveValue v)
+  => SnapshotName
+  -> Table IO k v b
+  -> IO () #-}
+createSnapshot :: forall m k v b.
+     ( IOLike m
+     , Common.Labellable (k, v, b)
+     , ResolveValue v
+     )
+  => SnapshotName
+  -> Table m k v b
+  -> m ()
+createSnapshot snap (Internal.Table' t) =
+    void $ Internal.createSnapshot (resolve (Proxy @v)) snap label Internal.SnapFullTable t
+  where
+    label = Internal.SnapshotLabel $ Common.makeSnapshotLabel (Proxy @(k, v, b))
+
+{-# SPECIALISE openSnapshot ::
+     (Common.Labellable (k, v, b), ResolveValue v)
+  => Session IO
+  -> Common.TableConfigOverride
+  -> SnapshotName
+  -> IO (Table IO k v b ) #-}
+openSnapshot :: forall m k v b.
+     ( IOLike m
+     , Common.Labellable (k, v, b)
+     , ResolveValue v
+     )
+  => Session m
+  -> Common.TableConfigOverride -- ^ Optional config override
+  -> SnapshotName
+  -> m (Table m k v b)
+openSnapshot (Internal.Session' sesh) override snap =
+    Internal.Table' <$!> Internal.openSnapshot sesh label Internal.SnapFullTable override snap (resolve (Proxy @v))
+  where
+    label = Internal.SnapshotLabel $ Common.makeSnapshotLabel (Proxy @(k, v, b))
+
+{-------------------------------------------------------------------------------
+  Mutiple writable tables
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE duplicate ::
+     Table IO k v b
+  -> IO (Table IO k v b) #-}
+duplicate ::
+     IOLike m
+  => Table m k v b
+  -> m (Table m k v b)
+duplicate (Internal.Table' t) = Internal.Table' <$!> Internal.duplicate t
+
+{-------------------------------------------------------------------------------
+  Table union
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE union ::
+     ResolveValue v
+  => Table IO k v b
+  -> Table IO k v b
+  -> IO (Table IO k v b) #-}
+union :: forall m k v b.
+     ( IOLike m
+     , ResolveValue v
+     )
+  => Table m k v b
+  -> Table m k v b
+  -> m (Table m k v b)
+union = error "union: not yet implemented" $ union @m @k @v @b
+
+{-------------------------------------------------------------------------------
+  Monoidal value resolution
+-------------------------------------------------------------------------------}
+
+resolve ::  forall v. ResolveValue v => Proxy v -> Internal.ResolveSerialisedValue
+resolve = coerce . resolveValue
+
+-- | Newtype wrapper for use with deriving via, which resolves values as a
+-- 'Data.Semigroup.First' semigroup.
+--
+-- The instance makes 'Insert's isomorphic to 'Mupsert's. This can be a sensible
+-- default if there is no intent to use 'Mupsert's.
```

Comment:
I'd rather say that the instance makes mupserts behave like inserts. So, the other way around, and without talking about isomorphisms, which we don't define in this context.

And I would explicitly mention that this avoids the need of providing a custom instance.